src/errors/either.ts

```diff
@@ -24,3 +24,15 @@ export type DefiniteEither<T, U> = {
   error?: T
   result: U
 }
+
+export const isFailure = <T, U>(e: Either<T, U>): e is Left<T> => {
+  return e.error !== undefined
+}
+
+export const isSuccess = <T, U>(e: Either<T, U>): e is Right<U> => {
+  return e.result !== undefined
+}
+
+export const failure = <T>(error: T): Left<T> => ({ error })
+
+export const success = <U>(result: U): Right<U> => ({ result })
```

Comment:
To make the return more explicit as Either constructor, wondering if could make sense to do something like

```
export const failure = <T>(error: T): Either<T, never> => ({ error })

export const success = <U>(result: U): Either<never, U> => ({ result })
```

Reply:
That would actually be a less-strict type as a result. It guarantees that wrong part is never, but it doesn't do any guarantees about the correct part - it is still `type generic | underfined`

Reply:
[There is not practical difference ](https://www.typescriptlang.org/play?#code/C4TwDgpgBAMhBmwA8AVAfFAvFA3gKCiggCdiB7YgLihQKmIgGcBXAG2AH5qA7CANxJ4AvnjyhIUAEoBLAOYALZAFUM2fIRLliXKLwHE6DFu2pLho8dACi04PJKoANFBVYoAOTLd3bVgEMAI1YIJDhEVAwAHyk5RSQVNFF4Zm4AY2BpLygWVNSmRgB5Xih4tAAKIzZgUwBKahs7Bz0SZ1d1eghgZmJuXA7jYCgRETxktIysnLzGRhQAdzISlQqmKtrqGQVlDHaGLp6+yvYh8zxUr0ZBnCPq-qqARiE3KfyiiDKcMHIwagByPj8rF+QhqZwuZGCADpWGRZCsBvdIV8yGBQWDuJdcDdqDcAExPbAvGbzMgfZE-KD-QHA0HnDEQiDQ2Hwqq4pHfGpAA)

Reply:
I understand, but suggested change achieves a different result compared to the original code

Reply:
yeah, the suggested change doesn't fit our needs, I am wondering if there is a way to return Either which make the property not being undefined

Reply:
Discussed on Slack, resolving conversation